files.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/custom-script-extension-linux/blobutil"
+	"github.com/Azure/custom-script-extension-linux/download"
+	"github.com/Azure/custom-script-extension-linux/preprocess"
+	"github.com/pkg/errors"
+)
+
+// downloadAndProcessURL downloads using the specified downloader and saves it to the
+// specified existing directory, which must be the path to the saved file. Then
+// it post-processes file based on heuristics.
+func downloadAndProcessURL(url, downloadDir, storageAccountName, storageAccountKey string) error {
+	fn, err := urlToFileName(url)
+	if err != nil {
+		return err
+	}
+
+	dl, err := getDownloader(
+		url,
+		storageAccountName,
+		storageAccountKey)
+	if err != nil {
+		return err
+	}
+
+	fp := filepath.Join(downloadDir, fn)
+	const mode = 0500 // we assume users download scripts to execute
+	if _, err := download.SaveTo(dl, fp, mode); err != nil {
+		return err
+	}
+
+	err = postProcessFile(fp)
+	return errors.Wrapf(err, "failed to post-process '%s'", fn)
+}
+
+// getDownloader returns a downloader for the given URL based on whether the
+// storage credentials are empty or not.
+func getDownloader(fileURL string, storageAccountName, storageAccountKey string) (
+	download.Downloader, error) {
+	if storageAccountName != "" && storageAccountKey != "" {
+		blob, err := blobutil.ParseBlobURL(fileURL)
+		if err != nil {
+			return nil, err
+		}
+		return download.NewBlobDownload(
+			storageAccountName,
+			storageAccountKey,
+			blob), nil
+	}
+	return download.NewURLDownload(fileURL), nil
+}
+
+// urlToFileName parses given URL and returns the section after the last slash
+// character of the path segment to be used as a file name. If a value is not
+// found, an error is returned.
+func urlToFileName(fileURL string) (string, error) {
```

Comment:
pretty much. we don't need the *url.URL anywhere else at all. Also `url.Parse` does not mean that the passed url is a full URL, it can be empty string, a relative url (e.g. `foo`) or a full url. We don't know for sure if the url is full (i.e. fetchable) until we attempt to download the file via http.Client.Get().